lib/evaluate/integer.h

```diff
@@ -0,0 +1,853 @@
+      if constexpr (partBits < 64) {
+        n = (n >> partBits) | signExtension;
+      } else {
+        n = signExtension;
+      }
+    }
+    SetLEPart(parts - 1, n);
+  }
+
+  constexpr Integer &operator=(const Integer &) = default;
+
+  // Left-justified mask (e.g., MASKL(1) has only its sign bit set)
+  static constexpr Integer MASKL(int places) {
+    if (places <= 0) {
+      return {};
+    } else if (places >= bits) {
+      return MASKR(bits);
+    } else {
+      return MASKR(bits - places).NOT();
+    }
+  }
+
+  // Right-justified mask (e.g., MASKR(1) == 1, MASKR(2) == 3, &c.)
+  static constexpr Integer MASKR(int places) {
+    Integer result{nullptr};
+    int j{0};
+    for (; j + 1 < parts && places >= partBits; ++j, places -= partBits) {
+      result.LEPart(j) = partMask;
+    }
+    if (places > 0) {
+      if (j + 1 < parts) {
+        result.LEPart(j++) = partMask >> (partBits - places);
+      } else if (j + 1 == parts) {
+        if (places >= topPartBits) {
+          result.LEPart(j++) = topPartMask;
+        } else {
+          result.LEPart(j++) = topPartMask >> (topPartBits - places);
+        }
+      }
+    }
+    for (; j < parts; ++j) {
+      result.LEPart(j) = 0;
+    }
+    return result;
+  }
+
+  static constexpr ValueWithOverflow ReadUnsigned(
+      const char *&pp, std::uint64_t base = 10) {
+    Integer result;
+    bool overflow{false};
+    const char *p{pp};
+    while (*p == ' ' || *p == '\t') {
+      ++p;
+    }
+    if (*p == '+') {
+      ++p;
+    }
+    Integer radix{base};
+    for (; std::uint64_t digit = *p; ++p) {
+      if (digit >= '0' && digit < '0' + base) {
+        digit -= '0';
+      } else if (base > 10 && digit >= 'A' && digit < 'A' + base - 10) {
+        digit -= 'A' - 10;
+      } else if (base > 10 && digit >= 'a' && digit < 'a' + base - 10) {
+        digit -= 'a' - 10;
+      } else {
+        break;
+      }
+      Product shifted{result.MultiplyUnsigned(radix)};
+      overflow |= !shifted.upper.IsZero();
+      ValueWithCarry next{shifted.lower.AddUnsigned(Integer{digit})};
+      overflow |= next.carry;
+      result = next.value;
+    }
+    pp = p;
+    return {result, overflow};
+  }
+
+  template<typename FROM>
+  static constexpr ValueWithOverflow ConvertUnsigned(const FROM &that) {
+    std::uint64_t field{that.ToUInt64()};
+    ValueWithOverflow result{field, false};
+    if constexpr (bits < 64) {
+      if ((field >> bits) != 0) {
+        result.overflow = true;
+        return result;
+      }
+    }
+    for (int j{64}; j < that.bits; j += 64) {
+      field = that.SHIFTR(j).ToUInt64();
+      if (bits <= j) {
+        if (field != 0) {
+          result.overflow = true;
+          break;
+        }
+      } else {
+        result.value = result.value.IOR(Integer{field}.SHIFTL(j));
+        if (bits < j + 64 && (field >> (bits - j)) != 0) {
+          result.overflow = true;
+          break;
+        }
+      }
+    }
+    return result;
+  }
```

Comment:
Here is a more concise body for ValueWithOverflow. Whether it is clearer, I'm not sure. I don't like the `break`s in the original.
```
    std::uint64_t field{that.ToUInt64()};
    ValueWithOverflow result{field, false};
    if constexpr (bits < 64) {
      result.overflow = (field >> bits) != 0;
    }
    for (int j{64}; j < that.bits && !result.overflow; j += 64) {
      field = that.SHIFTR(j).ToUInt64();
      if (bits <= j) {
        result.overflow = field != 0;
      } else {
        result.value = result.value.IOR(Integer{field}.SHIFTL(j));
        result.overflow = bits < j + 64 && (field >> (bits - j)) != 0;
      }
    }
    return result;
```

Reply:
Thanks.